### ISSUE TITLE
What happened to the first promise in "promise-it-wont-hurt" / "do-some-work"

### ISSUE THREAD

**Issue by OllieJones** [nodeschool/discussions#800]
The accepted solution to "do-some-work" contains this...

     HTTP.read("http://localhost:7000")
      .then(function (pk) {
            return HTTP.read("http://localhost:7001/" + pk)
       })
     .then( ...  more stuff ...)
     .done();

The second `return HTTP...`, the one in the function, looks like it instantiates another `promise`, and replaces the original one in the `then` chain.  Is that right? If so, what happens to the first `promise` object? Does it get `.done();` at any point?

(I must be missing something; I'm a C++ retread and paranoid about memory leaks).

**Comment by martinheidegger** [comment]
@OllieJones Thank you for asking a good question (that I have not thought of! :+1: )

In my understanding the Promise chaining `.then` means that everything within the braces will be called *after* the promise finished its work (or is .done).

In pseudocode:

```JavaScript

function then(something) {
   var newPromise = makePromise(something)
   this.on("done", newPromise.start.bind(newPromise))
   return newPromise
}
```

I hope that eases your mind ;)

**Comment by davecocoa** [comment]
In the spec, this is referred to as the [promise resolution procedure](https://promisesaplus.com/#point-45).

I'll modify your example a little and then try to explain what's going on.

```js
var prom1 = HTTP.read("http://localhost:7000");
var prom2 = prom1.then(function (pk) {
        var prom3 = HTTP.read("http://localhost:7001/" + pk);
        return prom3;
   });
var prom4 = prom2.then( ...  more stuff ...);
 prom4.done();
```

So I've deliberately named every promise in your example so that it's easier to talk about.

I believe you're interested in how `prom2` and `prom3` end up having roughly the same meaning. As we know, for promises returned from `then`, if the handler (`function(pk)...`) returns a value, the new promise is resolved with that value. However, if the handler (`function(pk)...`) returns a promise, the new promise "adopts its state".

That is to say, `prom2` and `prom3` have the same state - they both end up fulfilled with the same value, but they are *not* the same object. The spec language is a little bit hard to follow but once you get the hang of it it's a useful reference for things like this.

**Comment by OllieJones** [comment]
Thank you Martin.
